extensions/s3/src/main/java/io/deephaven/extensions/s3/S3AsyncClientFactory.java

```diff
@@ -0,0 +1,144 @@
+
+    private static class HttpClientConfig {
+        private final int maxConcurrentRequests;
+        private final Duration connectionTimeout;
+
+        HttpClientConfig(final int maxConcurrentRequests, final Duration connectionTimeout) {
+            this.maxConcurrentRequests = maxConcurrentRequests;
+            this.connectionTimeout = connectionTimeout;
+        }
+
+        int maxConcurrentRequests() {
+            return maxConcurrentRequests;
+        }
+
+        Duration connectionTimeout() {
+            return connectionTimeout;
+        }
+
+        @Override
+        public int hashCode() {
+            int result = maxConcurrentRequests;
+            result = 31 * result + connectionTimeout.hashCode();
+            return result;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            final HttpClientConfig that = (HttpClientConfig) o;
+            return maxConcurrentRequests == that.maxConcurrentRequests
+                    && connectionTimeout.equals(that.connectionTimeout);
+        }
+    }
+
+    private static SdkAsyncHttpClient getOrBuildHttpClient(@NotNull final S3Instructions instructions) {
+        final HttpClientConfig config = new HttpClientConfig(instructions.maxConcurrentRequests(),
+                instructions.connectionTimeout());
+        return httpClientCache.computeIfAbsent(config, key -> AwsCrtAsyncHttpClient.builder()
+                .maxConcurrency(config.maxConcurrentRequests())
+                .connectionTimeout(config.connectionTimeout())
+                .build());
+    }
+
+    /**
+     * The following executor will be used to complete the futures returned by the async client. This is a shared
+     * executor across all clients with fixed number of threads. This pattern is inspired by the default executor used
+     * by the SDK ({@code SdkAsyncHttpClient#resolveAsyncFutureCompletionExecutor}).
+     *
+     */
+    private static Executor ensureAsyncFutureCompletionExecutor() {
+        if (futureCompletionExecutor == null) {
+            synchronized (S3AsyncClientFactory.class) {
+                if (futureCompletionExecutor == null) {
+                    futureCompletionExecutor = Executors.newFixedThreadPool(NUM_PROCESSORS, new ThreadFactoryBuilder()
+                            .threadNamePrefix("aws-async-future-completion").build());
+                }
+            }
+        }
+        return futureCompletionExecutor;
+    }
+
+    /**
+     * The following executor will be used to schedule tasks for the async client, such as timeouts and retries. This is
+     * a shared executor across all clients with fixed number of threads. This pattern is inspired by the default
+     * executor used by the SDK ({@code ClientOverrideConfiguration#resolveScheduledExecutorService}).
```

Comment:
```suggestion
     * executor used by the SDK ({@code ClientOverrideConfiguration#scheduledExecutorService}).
```
Or
`software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder#resolveScheduledExecutorService`?